js/lib/AlphabeticIndex.js

```diff
@@ -60,159 +60,291 @@ var Collator = require("./Collator.js");
-							options.onLoad(this);
-						}
-	    			}
-	    		})
-			});
+
+			if (options && typeof(options.onLoad) === 'function') {
+				options.onLoad(this);
+			}
 		})
 	});
 };
 
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._updateCollationMap = function(style) {
+	this.collationMap = this.collation[style].map;
+	this.flowBoundaries = this.collation[style].flowBoundaries;
+	this.indexUnits = this.collation[style].indexUnits;
+	if (this.collation[style].inherit) {
+		this.inherit = this.collation[style].inherit;
+	}
+}
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._init = function(collation) {
+	if (!collation) {
+		return;
+	}
+
+	this.flowBoundaries = new Array();
+	
+	if (this.style === 'standard') {
+		this.style = this.collation["default"];
+	}
+	
+	this.collationMap = collation[this.style].map;
+	this.flowBoundaries = collation[this.style].flowBoundaries;
+	this.indexUnits = collation[this.style].indexUnits;
+	if (collation[this.style].inherit) {
+		this.inherit = collation[this.style].inherit;	
+	}
+	
+	if (this.collationMap === undefined &&
+		typeof(collation[this.style]) === 'string') {
+		this._updateCollationMap(collation[this.style]);
+	}
+}
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._getKeyByValue = function(value) {
+	var i,label;
+
+	if (!value || (!ilib.isArray(value))) {
+		return "";
+	}
+
+	if (this.inheritStyle) {
+		this.collationMap = this.collation[this.inheritStyle].map;
+	}
+
+	for (i=0; i < this.indexUnits.length; i++) {
+		if (this.collationMap[this.indexUnits[i]][0]
+			=== value[0]) {
+			label = this.indexUnits[i];
+			break;
+		}
+	}
+
+	return label;
+};
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._normalizeHangul = function(value) {
+	if (!value) {
+		return "";
+	}
+	var source = new NormString(value);
+	var normString = source.normalize("nfkd");
+	var firstJamo;
+	var it = normString.charIterator();
+
+	firstJamo = it.next();
+	return firstJamo;
+};
+
+/**
+ * Return the locale used with this instance.
+ * @return {Locale} the Locale instance for this index
+ */
+AlphabeticIndex.prototype.getLocale = function() {
+	return this.locale;
+};
+
 /**
  * Add an element to the index. The element is added to the
  * appropriate bucket and sorted within that bucket according
  * to the collation for the locale set up within this index.
- * 
+ *
  * @param {String} element the element to add
- * @returns {String} the label of the bucket into which
+ * @returns {String|undefined} the label of the bucket into which
  * this element was added
  */
 AlphabeticIndex.prototype.addElement = function(element) {
+	if (typeof(element) !== 'string') {
+		return undefined;
+	}
+
+	var i;
+	var label = this.getBucket(element);
+	var newItem = true;
+	var itemSet = {};
+
+	itemSet.label = label;
+	itemSet.elements = [];
+
+	for (i = 0; i < this.index.length; i++) {
+		if (this.index[i].label === label) {
+			this.index[i].elements.push(element);
+			this.index[i].elements.sort();
```

Comment:
This needs to use the collation to sort, otherwise you are sorting in Unicode. Also, you do not need to sort every time an element is added. The elements should be sorted only when they are returned to the caller in getAllBuckets().